Feature 3 overlapping shenanigans

Firebrandt opened the pull request:
OK!
**Changes Made & Fitting into overall project **

- More or less fully implements the OVERLAP MAXIMIZATION FEATURE/USE CASE, enhancing functionality and letting us satisfy more of the project specs. Includes implementations for all relevant interfaces and their class-implementations. I refrain from listing them all for the sake of brevity, but a lot of additional functionality has been added, going from an essentially non-functional branch to a nearly-complete branch that's just awaiting final integration. Lots of methods implemented. I honestly don't remember them all.

-  Adds a couple converters to assist in the translation of my InputDialog viewModel types and regular TimetableModel Model classes. **This should make extension a little bit easier since future manipulation of those types is now, well, easier. **

-  My use case's ARCHITECTURE IS NOW CLEAN RELATIVE TO LAST PULL REQUEST, LITERALLY REFACTORED 6+ TIMES this should substantially increase maintainability. Notable changes include switching from a super flawed architecture where my View worked w/ entities pass in through initialization, that was then passed into controllers which then passed them into the interactor, to a proper Clean Architecture that makes use of dependency injection & dependency inversion to limit dependency, and that, well, doesn't break the dependency rule constantly but that instead "flows" in the proper Clean Architecture way.
    This naturally involved the creation of various interfaces to support dependency inversion/injection, and some substantial refactoring.

  -  Also, instead of passing raw entities, I make proper use of various Model and ViewModel classes to pass data between layers now.

  -  Adds basic test classes for non-trivial methods as well, helping ensure code correctness (or at least quick diagnosis of breaks). Implemented equals() methods for several model classes to support unit tests and to make writing them in future more convenient. This was also missing from the first pull request! Does not cover super large numbers of edge cases, but there could be time to add that later on.

  -  Adds substantial initialization/setup code allowing for some integration with other use cases (expect full integration to come later though). Still, increases functionality somewhat since we can move through a flow of sorts now. This was also missing from the first pull request. TThis brings us closer to full integration with a (mostly) complete application, and allows for manual testing of the use case, which wasn't possible before since the use case wasn't previously runnable.

  -  Also fixed my file structure, so it matches the pattern decided on by the group instead of shoving everything into random but convenient directories.

  -  Sidenote: Much completed branch code from earlier commits was seemingly pulled before :p. I know this PR size seems underwhelming but I've put a lot of work into the branch.

**Design Decisions I straight up don't remember all of these but here are a few:**

  -  Used static to implement the singleton design pattern for my Converter classes, since we would never reasonably want to initialize more than 1 of these classes. They're just method holders! These are (I think) an implementation of an Adapter

 -   Technically my Controller works like an instance of the Adapter design pattern.

  -   Decided to UNDO an Observer Design Pattern from an earlier commit that at one point connted my Controller to my Dialog. I, uh, kind of realized that I was being trigger happy and applying a design pattern in a situation where instead of making code more maintainable, actually makes the code more overcomplicated and harder to maintain.

   -  Implemented the Observer Design Pattern for real to listen to JD's Timetable Generating Interactor, since in that particular case I DO need an interactor to be notified precisely when he has some use-case relevant data ready to send. Implemented this via Java's built in Flow interface!

    **- Refactored original architecture quite a few times to make the architecture properly Clean.**

   - Although it's not perfect, I had the controller and interactor methods return values without actually using the returned values in the main code. This is because it makes testing a lot easier and simpler, and I wanted tests to be simpler so that they wouldn't be flaky. Javadoc has texts warning users NOT TO depend on the returned values because it likely would break our Clean Architecture.

**To whoever is reviewing this, please LMK if there is additional work or detail you would like me to elaborate on. I'm just having trouble remembering everything since I've been working on this pull request for so long.**

Firebrandt committed:
Wrote a draft of the TimetableModelUnpacker class.

Firebrandt committed:
Drafted some unit tests for the unpacker class. Will need to debug.

Firebrandt committed:
Added equals() method for Course, overrides the obj one.
Added equals() for Session and Timetable too, to support the unit tests (which otherwise break).
Equality works the same as a Python dataclass.

Added a SessionType field to TimetableModel to support conversion between the two.

Firebrandt committed:
Implemented Michael H's revisions. Added additional documentation and small formatting edits to code.

Also improved the function of reformatModelTime since we decided on a change in the Discord.

Also handled the IntelliJ warnings.

Firebrandt committed:
Made the switch case into an if because the autograder sucks.

YahyaElgabra committed:
implemented TimetablesSortInteractor, documented my code

next steps: integrate with mainmenu and timetableUI, flow, tests

YahyaElgabra committed:
Merge remote-tracking branch 'origin/feature-13-finishing-sorter' into feature-13-finishing-sorter

YahyaElgabra committed:
checking if github still hates enhanced for loops

YahyaElgabra committed:
Implemented flow publisher, still didn't integrate it with the rest of my use case

YahyaElgabra committed:
Merge branch 'main' into feature-13-finishing-sorter

# Conflicts:
#	src/main/java/screens/MainUI.java

YahyaElgabra committed:
no implementations yet, stuck on how to convert TimetableViewModel into TimetableModel

YahyaElgabra committed:
Merge branch 'feature-9-finish-main-menu-ui' into feature-13-finishing-sorter

# Conflicts:
#	src/main/java/blacklist_whitelist_use_case/frameworks_and_drivers/ConstraintsInputScreen.java
#	src/main/java/screens/MainUI.java

YahyaElgabra committed:
I will go to WAR with swing

WHY CANT IT JUST LET ME SWITCH SCREENS AAAAAAAAAAAAAAH

YahyaElgabra committed:
started doing tests but stopped cuz i need sleep
for mainmenu integration: ask michael about swing
for ttui integration: ask people about converting from timetableviewmodel down
for flow: done

Firebrandt committed:
DESIGN DECISION MADE:
- UNDID THE OBSERVER PATTERN SINCE I WAS RUNNING INTO A KIND OF OVERCOMPLICATING IT CODE SMELL. REFACTORGURU MENTIONS IT, BUT BASICALLY, THE PATTERN OVERCOMPLICATES THINGS AND ACTUALLY MAKES IT HARDER.

- More refactoring since I found out some clean architecture was wrong. Turns out I need to pass in ViewModel to the InputDialog, so I've made those too.

- Added some skeletal support for TimetableUI from Kai. Hopefully this is good!

Firebrandt committed:
Factored out another CA violation into yet another convert! !!!
!!! !!! :D

YahyaElgabra committed:
still implementing AllTimetablesController

YahyaElgabra committed:
still implementing AllTimetablesController

YahyaElgabra committed:
observer pattern stuff

Firebrandt committed:
Merge remote-tracking branch 'origin/8-feature-2-generating-with-course-list' into feature-3-overlapping-shenanigans

# Conflicts:
#	src/main/java/blacklist_whitelist_use_case/frameworks_and_drivers/ConstraintsInputScreen.java

Firebrandt committed:
Merge remote-tracking branch 'origin/feature-13-finishing-sorter' into feature-3-overlapping-shenanigans

# Conflicts:
#	build.gradle
#	src/main/java/display_timetable_use_case/interface_adapters/TimetableUI.java
#	src/main/java/display_timetable_use_case/interface_adapters/TimetableViewBlockModel.java
#	src/main/java/screens/MainUI.java

Firebrandt committed:
Wrote the bones of um, a way to get data by listening to JD. Also happened to grab some stuff from Yahya - thx to him for implementing that publisher <3.

Firebrandt committed:
Wrote the bones of um, a way to get data by listening to JD. Also happened to grab some stuff from Yahya - thx to him for implementing that publisher <3.

Firebrandt committed:
Made a kind of mechanism to wait for JD's thing to finish. Ready to try running it now... I guess?

Firebrandt committed:
Set up a draft for my initialization code.

Firebrandt committed:
Fixed a few issues with the draft initialization.

Firebrandt committed:
Fixed a few issues with the draft initialization.

Firebrandt committed:
Fixed a few issues with the draft initialization.

Firebrandt committed:
Commit just to save

Firebrandt committed:
Merge remote-tracking branch 'origin/8-feature-2-generating-with-course-list' into feature-3-overlapping-shenanigans

Firebrandt committed:
Fixed architecture problem with the Controller (output goes through a Presenter to the View now :) )

Firebrandt committed:
Fixed unit tests for one of the Converters.

Firebrandt committed:
Made unit tests for the other converter.

Firebrandt committed:
Fixed my interactor unit tests >:D

Firebrandt committed:
Made new unit tests. Doing super integration now.

Refactored to not kill the main frame (instead refactors it).

Firebrandt committed:
Fixed my file layout.

Firebrandt committed:
Added a new function to update UI to the InputDialog.

Firebrandt committed:
Refined the UI resetting stuff a little bit.

michaelhsj commented on src/main/java/overlap_crap_fix_locations_later/application_business/CalculateSectionHoursInteractor.java:
```diff
@@ -0,0 +1,26 @@
+package overlap_crap_fix_locations_later.application_business;
```
If this is your code, could you please rename the package to something more descriptive?

Firebrandt replied:
Done! My bad, forgot to fix this earlier. But this is why we do prs!

michaelhsj commented on src/main/java/overlap_crap_fix_locations_later/application_business/CalculateSectionHoursInteractor.java:
```diff
@@ -0,0 +1,26 @@
+ *  the same as how we find the overlapping timeTable in principle.
+ */
+public class CalculateSectionHoursInteractor implements SectionHoursInputBoundary {
+
```
If this is an overriding method, I would recommend adding an @Override, just to make sure you're really overriding the boundary's methods.

Firebrandt replied:
oops! Fixed!

michaelhsj commented on src/main/java/overlap_crap_fix_locations_later/application_business/OverlapGeneratedTimetableRelayInteractor.java:
```diff
@@ -0,0 +1,66 @@
+ * <p>
+ * I am assuming that this is subscribed to something that gives me a List<Timetable>.
+ */
+public class OverlapGeneratedTimetableRelayInteractor implements Flow.Subscriber<Object> {
```
Does this interactor also implement an input boundary of any type?

Firebrandt replied:
There's no need in this case, since essentially its only job is to listen for data from another interactor and pass it up the food chain to the UI. Note how its only methods come from implementing Flow.Subscriber<Object> - hence the name 'Relay interactor' 😇 . 

You could look at it as not really needing an Input boundary, or you could look at it as having Flow.Subscriber<Object> as its input boundary. But we probably don't need another one on top of that.

michaelhsj commented on src/main/java/overlap_crap_fix_locations_later/application_business/TimeTableMatchInteractor.java:
```diff
@@ -0,0 +1,105 @@
+        // Although this is terribly inefficient, note that we have at most, like... 10 courses. Of 10 sections each.
+        // Section's don't store a reference to their parent course, so this is actually necessary D:
+        // Compare every section of 2 courses...
+        for (CourseModel mainCourse : mainCourses) {
```
If we have time in the future, it would be good to extract a method or two so we don't have so much nesting going on in one function, but we'll live with this for now.

Firebrandt replied:
I'll add the TODO, because you're right about this, but the presentation is in 5 hours so

michaelhsj commented on src/main/java/overlap_crap_fix_locations_later/frameworks_and_drivers/OverlapInputDialog.form:
```diff
@@ -0,0 +1,93 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Make sure that the Markdown file notes the prerequisites to compile this IntelliJ UI.

Firebrandt replied:
I added the reqs to the README since I couldn't open the XML Properly in IntelliJ.

michaelhsj commented on src/test/java/overlap_maximization_use_case/TestCalculateSectionHoursInteractor.java:
```diff
@@ -0,0 +1,55 @@
+//package overlap_crap_fix_locations_later.tests;
```
Is there any reason why the tests here are commented out? If we can make them work before merging, that would be for the best.

Firebrandt replied:
I don't know, that might have been from somebody else. But they're uncommented and ostensibly working now!

michaelhsj commented on src/main/java/overlap_crap_fix_locations_later/application_business/TimetableMatchInputBoundary.java:
```diff
@@ -0,0 +1,14 @@
+package overlap_crap_fix_locations_later.application_business;
+
```
This interface could use a brief explanation of what its methods do.

Firebrandt replied:
Done! Added documentation with warnings to not use the return values.

michaelhsj reviewed (commented):
Responding to the changes made here: https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name/pull/86#issue-1484118368

Okay! So this is overall looking much, _much_ cleaner than before, so I'm loving the architecture. I've left some comments on the code that you can take a look at, but the long and short is as follows:

I would change the overlap package name to something more descriptive and technical. Right now it seems quite loose.

Your OverlapGeneratedTimetableRelayInteractor doesn't seem to implement any input boundaries. I'm assuming it's the case that only other interactors call it, in which case I don't think there's any problem. But if a controller ends up calling it, make sure it implements some InputBoundary.

Your TimeTableMatchInteractor has some good code, but calculateTimetableOverlap has a 5 levels of nesting, which is a little smelly. If we have some time in the future, we should plan on extracting some of the method to make it easier to read and extend. It's a priority, but the lowest on the list.

TimetableMatchInputBoundary could use a couple sentences of documentation.

The tests in TestCalculateSectionHoursInteractor are all commented out right now. Do these tests work?

Firebrandt committed:
Implement Michael H's revisions. Fixed warnings in my files.

Added a strings.properties file to store UI strings. Centralizing them in one place could be a good idea to increase UI maintainability (what if we need to translate this abominable application into a different language?)

Firebrandt committed:
Merge remote-tracking branch 'origin/main'

# Conflicts:
#	src/main/java/blacklist_whitelist_use_case/frameworks_and_drivers/ConstraintsInputScreen.java
#	src/main/java/display_timetable_use_case/frameworks_and_drivers/TimetableUI.java
#	src/main/java/display_timetable_use_case/frameworks_and_drivers/TimetableViewBlockModel.java
#	src/main/java/display_timetable_use_case/interface_adapters/DisplayTimetableController.java
#	src/main/java/edit_timetable_use_case/frameworks_and_drivers/AddSectionsMenu.java
#	src/main/java/fileio_use_case/frameworks_and_drivers/TimetableGateway.java
#	src/main/java/overlap_crap_fix_locations_later/OverlapInputDialog.java
#	src/main/java/screens/MainUI.java
#	src/main/java/timetables_sort_use_case/frameworks_and_drivers/AllTimetablesScreen.java
#	src/test/java/overlap_stuff_fix_layout_later/OverlapInputDialogTest.java

Firebrandt committed:
Refactored the equals methods to be a little less terrible and hopefully not break autograder.

Firebrandt committed:
Merge branch 'main' into 66-TimetableModelConverter

# Conflicts:
#	src/main/java/entities/Block.java
#	src/main/java/entities/Section.java
#	src/main/java/entities/Timetable.java
#	src/main/java/entities/TimetableCourse.java
#	src/main/java/retrieve_timetable_use_case/RetrieveTimetableInteractor.java
#	src/main/java/retrieve_timetable_use_case/TimetableModel.java
#	src/main/java/retrieve_timetable_use_case/application_business/EntityConverter.java
#	src/main/java/retrieve_timetable_use_case/interface_adapters/TimetableModelConverter.java

Firebrandt committed:
Merge pull request #68 from CSC207-2022F-UofT/66-TimetableModelConverter

66 timetable model converter is go!

Firebrandt committed:
Resolve conflicts with main, also implement the Publisher Kai's case needs supposedly.

Switches the updateTimetable() call used for Kai to the proper one as well.

Prep for merger!!!

Firebrandt committed:
Merge branch '79-final-integration' into feature-3-overlapping-shenanigans

michaelhsj reviewed (approved):
Looks good! See the other Feature 3 branch, everything here is clear.
